contracts/contracts/strategies/BaseConvexMetaStrategy.sol

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.8.0;
+
+/**
+ * @title Curve Convex Strategy
+ * @notice Investment strategy for investing stablecoins via Curve 3Pool
+ * @author Origin Protocol Inc
+ */
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { IRewardStaking } from "./IRewardStaking.sol";
+import { ICurvePool } from "./ICurvePool.sol";
+import { ICurveMetaPool } from "./ICurveMetaPool.sol";
+import { IERC20, BaseCurveStrategy } from "./BaseCurveStrategy.sol";
+import { StableMath } from "../utils/StableMath.sol";
+import { Helpers } from "../utils/Helpers.sol";
+
+abstract contract BaseConvexMetaStrategy is BaseCurveStrategy {
+    using StableMath for uint256;
+    using SafeERC20 for IERC20;
+
+    address internal cvxDepositorAddress;
+    address internal cvxRewardStakerAddress;
+    uint256 internal cvxDepositorPTokenId;
+    ICurveMetaPool internal metapool;
+    IERC20 internal metapoolMainToken;
+    IERC20 internal metapoolLPToken;
+    // Ordered list of metapool assets
+    address[] internal metapoolAssets;
+
+    /**
+     * Initializer for setting up strategy internal state. This overrides the
+     * InitializableAbstractStrategy initializer as Curve strategies don't fit
+     * well within that abstraction.
+     * @param _rewardTokenAddresses Address of CRV & CVX
+     * @param _assets Addresses of supported assets. MUST be passed in the same
+     *                order as returned by coins on the pool contract, i.e.
+     *                DAI, USDC, USDT
+     * @param _pTokens Platform Token corresponding addresses
+     * @param _initAddresses Various addresses containing the following:
+     *  - _platformAddress Address of the Curve 3pool
+     *  - _vaultAddress Address of the vault
+     *  - _cvxDepositorAddress Address of the Convex depositor(AKA booster) for this pool
+     *  - _metapoolAddress Address of the Curve MetaPool
+     *  - _metapoolMainToken Address of Main metapool token
+     *  - _cvxRewardStakerAddress Address of the CVX rewards staker
+     * @param _cvxDepositorPTokenId Pid of the pool referred to by Depositor and staker
+     */
+    function initialize(
+        address[] calldata _rewardTokenAddresses, // CRV + CVX
+        address[] calldata _assets,
+        address[] calldata _pTokens,
+        /**
+         * in the following order: _platformAddress(3Pool address), _vaultAddress, _cvxDepositorAddress
+         * _metapoolAddress, _metapoolMainToken, _cvxRewardStakerAddress, _metapoolLPToken
+         */
+        address[] calldata _initAddresses,
```

Comment:
Was using individual arguments hitting the stack too deep error?

I'm wondering about the tradeoffs here between an array vs a struct for `_initAddresses`. Seems a little prone to confusion to use an array with each element being a completely different kind of thing, and the only visible difference being at the documentation level.